### DISCUSSION TITLE
Network emulation so you can run servers virtually in daedalOS

### DISCUSSION THREAD

**Discussion by Jayx239** [DustinBrett/daedalOS#476] [Feature Requests]
Not sure if this is already supported or not. But I thought it would be cool to be able to spin up a virtual server in daedalOS such that you would be able to develop and run servers in the virtual environment, and make calls to "virtual" localhost in the daedalOS instance. I'm currently diving into the code to get an idea of how everything is done, and it's super impressive. Would really enjoy working on this feature in my free time if this is something you'd think would be a good addition.

**Comment by DustinBrett** [comment]
Thanks for the suggestion! It is indeed something I have considered and hopefully one day I will get into that. I think I could use something like [Mock Service Worker](https://mswjs.io/) to make it happen, or at least Service Workers in general which allow me to handle all requests to internal/external sources.

**Comment by bitnom** [comment]
IMO creating a networking "stack" should be avoided at all costs since the browser already has one. CORS is the only annoyance but a service worker + ws bridge will cure all. You would just miss out on networking some of those fancy emulators unless they bundle their own network adapter.

It sounds like a nightmare to create such a stack from nothing; however, I just found this: https://github.com/chipmk/tcpip.js

edit: also found this: https://github.com/cbdevnet/websocksy
